pkg/vault/contracts/BalancerContractRegistry.sol

```diff
@@ -0,0 +1,278 @@
+     * );
+     * addOrUpdateBalancerContractAlias('WeightedPool', 0x9FC3da866e7DF3a1c57adE1a97c9f00a70f010c8);
+     *
+     * At that point,
+     * getBalancerContract(ContractType.POOL_FACTORY, '20241205-v3-weighted-pool') returns active=false,
+     * isActiveBalancerContract(ContractType.POOL_FACTORY, 0x201efd508c8DfE9DE1a13c2452863A78CB2a86Cc) returns false,
+     * getBalancerContract(ContractType.POOL_FACTORY, 'WeightedPool') returns the v2 address (and active=true).
+     */
+
+    /// @inheritdoc IBalancerContractRegistry
+    function registerBalancerContract(
+        ContractType contractType,
+        string memory contractName,
+        address contractAddress
+    ) external authenticate {
+        // Ensure arguments are valid.
+        if (contractAddress == address(0)) {
+            revert ZeroContractAddress();
+        }
+
+        if (bytes(contractName).length == 0) {
+            revert InvalidContractName();
+        }
+
+        // Ensure address isn't already in use.
+        ContractInfo memory info = _contractInfo[contractAddress];
+        if (info.isRegistered) {
+            revert AddressAlreadyRegistered(info.contractType, contractAddress);
+        }
+
+        // Ensure name isn't already in use (including as an alias).
+        bytes32 contractId = _getContractId(contractName);
+        address existingContract = _contractRegistry[contractId];
+
+        if (existingContract == address(0)) {
+            existingContract = _contractAliases[contractId];
+        }
+
+        if (existingContract != address(0)) {
+            info = _contractInfo[existingContract];
+
+            revert ContractAlreadyRegistered(info.contractType, contractName);
+        }
+
+        // Store the address in the registry, under the unique name.
+        _contractRegistry[contractId] = contractAddress;
+
+        // Record the address as active. The `isActive` flag enables differentiating between unregistered and deprecated
+        // addresses.
+        _contractInfo[contractAddress] = ContractInfo({
+            contractType: contractType,
+            isRegistered: true,
+            isActive: true
+        });
+
+        emit BalancerContractRegistered(contractType, contractName, contractAddress);
+    }
+
+    /// @inheritdoc IBalancerContractRegistry
+    function deregisterBalancerContract(string memory contractName) external authenticate {
+        // Ensure the name is registered
+        bytes32 contractId = _getContractId(contractName);
+        address contractAddress = _contractRegistry[contractId];
+
+        if (contractAddress == address(0)) {
+            revert ContractNotRegistered();
+        }
+
+        ContractInfo memory info = _contractInfo[contractAddress];
+        // This should be impossible: the registry and info mappings must be in sync.
+        require(info.isRegistered, "Registry entry with no contract info");
```

Comment:
i'd stick to `if` / `revert`, unless there's consensus to change the compiler version for everything and start using `require` with custom errors from now on, but that's a larger change.

I think we'd have used `require` if we had it from the start, but at the point the compiler started supporting it it was already too late to update the vault.

Reply:
Do we have a distinction between errors that a user can cause (which would revert with custom errors), and errors that are "invariant violations" that, if triggered, mean something unforeseen has happened, making the state internally inconsistent?

If not, could just remove this. If we still want a custom error here, it would be something like "InconsistentState"

Reply:
Not really.
We've used `require` inside tests because it's convenient to differentiate from assertions, and also because we don't really care about error codes inside tests. But elsewhere we've always used `if / revert` (except in gyro pools, which are a port, and using `if / revert` was an abomination, so we upgraded the contract version).

I'd leave the check, but use `if / revert` instead to continue with the usual pattern.